tests/spell_check3.k

```diff
@@ -0,0 +1,318 @@
+    role SpellChecker {
+        public void advanceToNextWord() {
+            if(isFinished()) return;
+
+            // Advance past current word
+            while(isAlphabetic(currentCharacter())) {
+                setPosition(currentPosition() + 1);
+                if(isFinished()) return;
+            }
+
+            // Advance past non-alphabetic characters
+            while(!isAlphabetic(currentCharacter())) {
+                setPosition(currentPosition() + 1);
+                if(isFinished()) return;
+            }
+
+            Dictionary.checkCurrentWord();
+        }
+
+        public String currentWord() {
+            if(isFinished()) return "";
+
+            String output = "";
+            int pos = currentPosition().clone;
+
+            while(isAlphabetic(Text.characterAt(pos))) {
+                output = output + Text.characterAt(pos++);
+                // NOTE: Cannot call Text.getText().length() here.
+                if(pos >= Text.length()) return output;
+            }
+
+            return output;
+        }
+
+        private String currentCharacter() {
+            return Text.characterAt(currentPosition());
+        }
+
+        private boolean isFinished() {
+            return currentPosition() >= Text.length();
+        }
+
+        private boolean isAlphabetic(String character) {
+            assert(character.length() == 1, "character string length not equal to one.");
+            return character == "-" || 
+                   (character >= "a" && character <= "z") ||
+                   (character >= "A" && character <= "Z");
+        }
+    } requires {
+        int currentPosition() const;
+        void setPosition(int pos);
+    }
+
+    /////////////////////////////////////////
+    role Text {
+        public String characterAt(int pos) {
+            return getText().substring(pos, pos+1);
+        }
+        public int length() {
+            return getText().length();
+        }
+    } requires {
+        String getText() const;
+        void splice(int startIndex, int endIndex, String text);
+    }
+
+    /////////////////////////////////////////
+    stageprop Dictionary {
+        public void checkCurrentWord() {
+            String word = SpellChecker.currentWord();
+            List<String> suggestions = suggestSpelling(word);
+
+            if(suggestions.size() > 0) {
+                int pos = SpellChecker.currentPosition();
+
+                // TODO: User input and selection
+                Text.splice(pos, pos + word.length(), "[" + suggestions.get(0) + "]");
+            }
+
+            SpellChecker.advanceToNextWord();
+        }
+    } requires {
+        List<String> suggestSpelling(String word) const;
+    }
+
+    /////////////////////////////////////////
+
+    public SpellCheck(Text text) {
+        Text = text;
+        SpellChecker = this;
+        Dictionary = new SimpleDictionary();
+    }
+
+    public void start() {
+        setPosition(0);
```

Comment:
I am a bit concerned that the current restriction in trygve from 
accessing Context members from role methods may cause programmers to 
nearly always make the context play a role, merely to make context 
properties accessible - whether that matches the mental model or not. In 
this case, I suppose it does make sense for "SpellChecker" to be a role, 
but I think there are other mental models that would be valid for this 
but difficult to accomplish given that restriction.

We may want to revisit the possibility of accessing Context members from 
role methods as long as there isn't a name clash, similar to how 
closures work in some other languages.

On 1/18/16 11:38 AM, Andreas wrote:
>
> In tests/spell_check3.k 
> <https://github.com/jcoplien/trygve/pull/50#discussion_r50019126>:
>
> > +            SpellChecker.advanceToNextWord();
> > +        }
> > +    } requires {
> > +        List<String> suggestSpelling(String word) const;
> > +    }
> > +
> > +    /////////////////////////////////////////
> > +
> > +    public SpellCheck(Text text) {
> > +        Text = text;
> > +        SpellChecker = this;
> > +        Dictionary = new SimpleDictionary();
> > +    }
> > +
> > +    public void start() {
> > +        setPosition(0);
>
> Could be, but a text document doesn't have the current position of the 
> spell checker. The |SpellChecker| RoleObjectContract is:
>
>      int  currentPosition() const;
>      void  setPosition(int  pos);
>
> For me, the role |Text| is simply a text buffer. I can understand it 
> could have RoleMethods for setting and getting the current position, 
> but where should the current position be stored?
>
> —
> Reply to this email directly or view it on GitHub 
> <https://github.com/jcoplien/trygve/pull/50/files#r50019126>.
>